819 - refactor use_files.R before adding replace arg

ilyaZar opened the pull request:
To fix #819 , there is the possibility to refactor the internals of `use_external_XXX` type functions in  `use_files.R ` **_prior to the addition of a new replace arg_**.

This may or may not be useful; I think it could be, hence the draft PR with some benefits being to reduce duplication

It's easier to add the 'replace'-argument in only one place then (see last commit to this PR).

Essentially all `use_external_xxx` funcs in `use_files.R` do the same, so this PR tries to encapsulate identical behaviour into smaller maintainable pieces. 

Of course this proposal comes along with quite some work and probably there should be some tests added as well... happy do get on this though!

ilyaZar committed:
refactor: refactor use_files.R before adding arg

Aim:
- reduce duplication
- make long term maintanence easier
- ease the later addition of 'replace'-argument as this needs then
to be done only in one place

How:
- encapsulate internals of use_external_js_file to separate funcs
- re-use separate funcs to reduce duplication in use_external_css_file

Possible To-Dos:
- re-use separate funcs to reduce duplication in
  use_external_html_template
- re-use separate funcs to reduce duplication in use_external_file

ilyaZar committed:
refactor: unify internal of use_external_html_template

- has same structure as use_external_{js,css}_file()
- add check_url_valid

ilyaZar committed:
refactor: update use_external_file as far possible

- no dance for use_external_file
- no url check for use_external
- no new-file creation for use_external

Otherwise, same internal structure as other use_external_XXX funcs

VincentGuyader commented:
HI, 

i prefer `overwrite = TRUE`, instead of `replace = TRUE`.
ok for you @ColinFay ?

ilyaZar committed:
style: restyle with grkstyler

ilyaZar committed:
style: remove redundant dev comments

ilyaZar commented:
## Update

Update this PR-branch to be up-to-date with current `dev`:

1. change `replace`-argument name to `overwrite` 
2. use the following style https://github.com/ThinkR-open/golem/commit/7ba820b3693753d99a58fb309e4278db7e326194#diff-c03aab5bb43f6562efff9eeb973b449f684e38dd79aed2b839008ffaeee45a8eR183-R194 for `get_new_dir()` helper function to throw an error
3. use `check_name_length_is_one` instead of `check_name_length()` in `get_new_name()`-helper function
4. set name default argument value for `name` to `NULL`

1-4 are updates made in https://github.com/ThinkR-open/golem/commit/7ba820b3693753d99a58fb309e4278db7e326194#diff-c03aab5bb43f6562efff9eeb973b449f684e38dd79aed2b839008ffaeee45a8e

ilyaZar committed:
fix: handle default name=NULL

default value can be NULL and NULL can be passed as argument
    -> check for both cases
    -> use basename of the URL correctly in this case

ilyaZar committed:
tests: fix test failiure

`file_path_sans_ext(name)` gets rid of the extension
this must be handled inside tests correctly

ilyaZar committed:
docs: update man-info to pass CI

ColinFay commented:
Hey, 

Thanks a lot for your contribution (as usual, you rock 🤘). 

I've worked on a full refactoring of the use_* family that you can find here : https://github.com/ThinkR-open/golem/pull/1170 

I've gone way deeper into factoring the function so that it's more consistent, and it will make it easier to add new use_* functions if we wish to. 

Closing this one, but hanks again 🙏
